In [50]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
sns.set_style('whitegrid')
%matplotlib inline

import chart_studio.plotly as py

In [51]:
import plotly.graph_objs as go 
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot

In [52]:
import cufflinks as cf

In [53]:
# For Notebooks
init_notebook_mode(connected=True)

In [54]:
# For offline use
cf.go_offline()

# COVID-19

## Data Analysis and Visualization 

### Overview

This project is based on a data set with daily information about: 
        - number of new cases of Covid-19 registered
        - number of deaths registered due to Covid-19

for every country in the world, until 29/07/2020.

On the Data Frame there is also each country's population size as of 2019, as well as the cumulative number for 14  days of COVID-19 cases per 100000 people.

### Analysis

On this notebook we look at the data from three different prespectives:
    - Portugal
    - World
    - Europe

There will be analysis and visualizations about the data's behavior through the period of time mentioned above.

### Note

With this analysis I am trying to learn more about data science and how to play with pandas and visualization.
There is a long way to go but trying and getting experience never hurt. I am using this data set to develop my skills as an aspiring data scientist since the best way to learn is to practice, not so much to bring new information to the table. There are so many amazing analysis out there about the pandemic's numbers and I am not trying to compare mine to those by any means.

In [55]:
covid_world = pd.read_csv('covid19-world')

covid_world.head()

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0  29/07/2020   29      7  2020    103       1             Afghanistan    AF   
1  28/07/2020   28      7  2020    105       1             Afghanistan    AF   
2  27/07/2020   27      7  2020    106      10             Afghanistan    AF   
3  26/07/2020   26      7  2020    121      13             Afghanistan    AF   
4  25/07/2020   25      7  2020    108      35             Afghanistan    AF   

  countryterritoryCode  popData2019 continentExp  \
0                  AFG   38041757.0         Asia   
1                  AFG   38041757.0         Asia   
2                  AFG   38041757.0         Asia   
3                  AFG   38041757.0         Asia   
4                  AFG   38041757.0         Asia   

   Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                           4.550263           
1                                           5.028685           
2                                           4.763187           
3                                           4.484546           
4                                           4.389913

## Portugal

In [56]:
# Create a new data frame only with information for Portugal

data_portugal = pd.DataFrame()
data_portugal = covid_world[covid_world['countriesAndTerritories']=='Portugal']

Total cases since 03/03/2020 (until 29/07/2020)

In [57]:
data_portugal['cases'].sum()

50410

Total number of deaths since 03/03/2020 (until 29/07/2020)

In [58]:
data_portugal['deaths'].sum()

1722

In [59]:
data_portugal.reset_index(inplace=True)
data_portugal.head()

index     dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0  24369  29/07/2020   29      7  2020    111       3                Portugal   
1  24370  28/07/2020   28      7  2020    135       2                Portugal   
2  24371  27/07/2020   27      7  2020    209       1                Portugal   
3  24372  26/07/2020   26      7  2020    263       4                Portugal   
4  24373  25/07/2020   25      7  2020    313       7                Portugal   

  geoId countryterritoryCode  popData2019 continentExp  \
0    PT                  PRT   10276617.0       Europe   
1    PT                  PRT   10276617.0       Europe   
2    PT                  PRT   10276617.0       Europe   
3    PT                  PRT   10276617.0       Europe   
4    PT                  PRT   10276617.0       Europe   

   Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                          32.685854           
1                                          33.873015           
2                                          35.536987           
3                                          36.334915           
4                                          39.049816

Total number of new cases per month (from march until july)

In [60]:
data_portugal.groupby('month')['cases'].sum()

month
3     6408
4    18284
5     7511
6     9709
7     8498
Name: cases, dtype: int64

Evolution of new cases per month (plot)

In [61]:
data_portugal.groupby('month')['cases'].sum().iplot(kind='bar', title='New Cases per Month')

Number of deaths per month (from march until july)

In [62]:
data_portugal.groupby('month')['deaths'].sum()

month
3    140
4    849
5    407
6    172
7    154
Name: deaths, dtype: int64

Evolution of number of deaths per month (plot)

In [63]:
data_portugal.groupby('month')['deaths'].sum().iplot(kind='bar')

Single day with the most new cases registered

In [64]:
data_portugal[data_portugal['cases'] == data_portugal['cases'].max()]

index     dateRep  day  month  year  cases  deaths  \
109  24478  11/04/2020   11      4  2020   1516      26   

    countriesAndTerritories geoId countryterritoryCode  popData2019  \
109                Portugal    PT                  PRT   10276617.0   

    continentExp  Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
109       Europe                                         109.024205

Single day with the most deaths registered

In [65]:
data_portugal[data_portugal['deaths'] == data_portugal['deaths'].max()]

index     dateRep  day  month  year  cases  deaths  \
95  24464  25/04/2020   25      4  2020    918      60   

   countriesAndTerritories geoId countryterritoryCode  popData2019  \
95                Portugal    PT                  PRT   10276617.0   

   continentExp  Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
95       Europe                                          75.890733

As we can see, the peak of new cases happened in april, with the single day with most new cases registered being 25/04/2020 with 918.

## World

Now we will focus our attention on the worldwide data.

Total registered cases of Covid-19 per country

In [66]:
covid_world.groupby('countriesAndTerritories')['cases'].sum().sort_values(ascending=False)

countriesAndTerritories
United_States_of_America             4351997
Brazil                               2483191
India                                1531669
Russia                                823515
South_Africa                          459761
Mexico                                402697
Peru                                  395005
Chile                                 349800
United_Kingdom                        300692
Iran                                  296273
Spain                                 280610
Pakistan                              276288
Saudi_Arabia                          270831
Colombia                              267385
Italy                                 246488
Bangladesh                            229185
Turkey                                227982
Germany                               206926
France                                183804
Argentina                             167403
Iraq                                  115332
Canada                         

Country with the highest number of new cases registered on a single day:

In [67]:
covid_world[covid_world['cases']==covid_world['cases'].max()]

dateRep  day  month  year  cases  deaths   countriesAndTerritories  \
31428  25/07/2020   25      7  2020  78427    1304  United_States_of_America   

      geoId countryterritoryCode  popData2019 continentExp  \
31428    US                  USA  329064917.0      America   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
31428                                         281.979619

Disclaimer: 

We conclude that the USA had the highest number of new cases on a single day (25-07-2020 with 78427 new cases) but we did not take into consideration the population.

A better indicator would be number of new cases / size of population



Plot the total cases per country on a world map

In [68]:
# Creating a new Data Frame with the total by country and with the correpsondent country code

total = pd.DataFrame(columns=['CountryCode', 'CountryName', 'TotalCases'])

total['CountryCode'] = list(covid_world.groupby('countriesAndTerritories')['countryterritoryCode'].unique())
total['CountryName'] = list(covid_world.groupby('countriesAndTerritories')['countriesAndTerritories'].unique())
total['TotalCases'] = list(covid_world.groupby('countriesAndTerritories')['cases'].sum())

total['CountryCode'] = total['CountryCode'].str.get(0)
total['CountryName'] = total['CountryName'].str.get(0)

total.head()

CountryCode  CountryName  TotalCases
0         AFG  Afghanistan       36471
1         ALB      Albania        4997
2         DZA      Algeria       27973
3         AND      Andorra         907
4         AGO       Angola         932

In [69]:
data = dict(type='choropleth',
           locations = total['CountryCode'],
           z = total['TotalCases'],
           colorbar = {'title': 'number of cases'},
        )

In [70]:
layout = dict(title = 'Total Number of COVID-19 Cases',
             geo = dict(showframe=False,
                       projection = {'type': 'mercator'},
                       )
             )

In [71]:
choromap = go.Figure(data = [data], layout = layout)
iplot(choromap)

Plot the total cases per country, divided by the population, on a world map

In [72]:
# Add a column to the total Data Frame called 'Population' with the population number for each country as of 2019

total['Population'] = list(covid_world.groupby('countriesAndTerritories')['popData2019'].unique())

total['Population'] = total['Population'].str.get(0)
total.head()

CountryCode  CountryName  TotalCases  Population
0         AFG  Afghanistan       36471  38041757.0
1         ALB      Albania        4997   2862427.0
2         DZA      Algeria       27973  43053054.0
3         AND      Andorra         907     76177.0
4         AGO       Angola         932  31825299.0

In [73]:
data = dict(type='choropleth',
           locations = total['CountryCode'],
           z = total['TotalCases']/total['Population'],
           colorbar = {'title': 'number of cases'},
        )

In [74]:
layout = dict(title = 'Total Number of COVID-19 Cases considering Population size',
             geo = dict(showframe=False,
                       projection = {'type': 'mercator'},
                       ),
             )

In [75]:
choromap = go.Figure(data = [data], layout = layout)
iplot(choromap)

## Europe

Now let's turn our attention to Europe only.

First create a new Data Frame only with european countries

In [76]:
europe = pd.DataFrame()
europe = covid_world[covid_world['continentExp']=='Europe']

europe.head()

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
202  29/07/2020   29      7  2020    117       4                 Albania   
203  28/07/2020   28      7  2020    117       6                 Albania   
204  27/07/2020   27      7  2020    126       4                 Albania   
205  26/07/2020   26      7  2020     67       6                 Albania   
206  25/07/2020   25      7  2020    104       5                 Albania   

    geoId countryterritoryCode  popData2019 continentExp  \
202    AL                  ALB    2862427.0       Europe   
203    AL                  ALB    2862427.0       Europe   
204    AL                  ALB    2862427.0       Europe   
205    AL                  ALB    2862427.0       Europe   
206    AL                  ALB    2862427.0       Europe   

     Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
202                                          46.464067           
203                                          45.730424           
204                                          45.730424           
205                                          44.228202           
206                                          45.136522

Total number of cases registered per country in Europe:

In [77]:
# Create a new Data Frame with the totals for the european countries

europe.groupby('countriesAndTerritories')['cases'].sum().sort_values(ascending=False)

countriesAndTerritories
Russia                    823515
United_Kingdom            300692
Spain                     280610
Italy                     246488
Germany                   206926
France                    183804
Sweden                     79494
Belarus                    67366
Ukraine                    66575
Belgium                    66569
Netherlands                53374
Portugal                   50410
Romania                    47053
Poland                     43904
Armenia                    37937
Switzerland                34522
Azerbaijan                 30858
Ireland                    25929
Serbia                     24520
Moldova                    23521
Austria                    20682
Czechia                    15799
Denmark                    13577
Bulgaria                   10871
Bosnia_and_Herzegovina     10766
North_Macedonia            10316
Norway                      9132
Kosovo                      7652
Finland                     7404
Luxembourg         

Country with the highest number of new cases of a single day:

In [78]:
europe[europe['cases']==europe['cases'].max()]

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
25071  18/07/2020   18      7  2020  12640     310                  Russia   

      geoId countryterritoryCode  popData2019 continentExp  \
25071    RU                  RUS  145872260.0       Europe   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
25071                                           66.87632

Disclaimer:

As with the worldwide analysis, we are not taking into consideration the size of population. Russia is the biggest country in Europe when it comes to population size.
Once again, a better indicator would be: number of new cases / population size

Analysing the single month with the highest number of new cases registered

In [79]:
europe.groupby(['countriesAndTerritories', 'month'])['cases'].sum().sort_values(ascending=False)

countriesAndTerritories  month
Russia                   5        297176
                         6        244581
                         7        182359
United_Kingdom           4        143080
Spain                    4        110916
United_Kingdom           5        107605
Spain                    3        104213
Italy                    4        101852
                         3        100851
Russia                   4         97563
Germany                  4         97206
France                   4         83892
Germany                  3         61856
France                   3         44493
United_Kingdom           6         39139
Belgium                  4         34454
Spain                    7         31339
Sweden                   6         30554
Italy                    5         29073
Belarus                  5         28477
Netherlands              4         27052
Spain                    5         24246
France                   5         23054
Ukraine                  7

With this Series we can conclude that:
    Russia had the highest number of new cases registered on a single month in may with a total of 297176 cases

## Conclusion

With this analysis it was possible to work with real data and explore many tools of the Pandas library as well as some Seaborn visualizations.

